lib/spack/spack/solver/asp.py

```diff
@@ -1670,8 +1670,9 @@ def variant_value(self, pkg, name, value):
         pkg_class = spack.repo.path.get_pkg_class(pkg)
 
         variant = self._specs[pkg].variants.get(name)
-        if variant:
-            # it's multi-valued
+        if variant and (type(variant) == spack.variant.SingleValuedVariant):
+            variant.value = value
+        elif variant:
```

Comment:
I believe these lines cover an inherent issue. As you pointed out to me in DMs we have these rules and facts:
```
variant_set("py-sip","module","PyQt5.sip") :- depends_on("py-pyqt5","py-sip"),node("py-pyqt5"),version_satisfies("py-pyqt5",":5.13.0").
variant_set("py-sip","module","PyQt5.sip") :- depends_on("py-pyqt5","py-sip"),node("py-pyqt5").
variant("py-sip","module").
variant_single_value("py-sip","module").
variant_default_value_from_package_py("py-sip","module","sip").
variant_possible_value("py-sip","module","sip").
```
I think we just miss a:
```
variant_possible_value("py-sip","module","PyQt5.sip").
```
and need no modifications in `asp.py` to set a value for a single valued variant. What happens is that these two rules:
```
1 {
  variant_value(Package, Variant, Value)
  : variant_possible_value(Package, Variant, Value)
} 1
 :- node(Package),
    variant(Package, Variant),
    variant_single_value(Package, Variant).

variant_value(Package, Variant, Value)
 :- node(Package),
    variant(Package, Variant),
    variant_set(Package, Variant, Value).
```
and the missing "possible value" make it such that we have two different values computed for a single valued variant.

Reply:
The difficulty here is that the `module` variant accepts any string as a value, so we should probably track the values mentioned during inspection of the packages, and emit them as possible values at the end.